drivers/adc/ad7091r8/ad7091r8.c

```diff
@@ -0,0 +1,578 @@
+		break;
+	case AD7091R8_HYSTERESIS:
+		reg = AD7091R8_REG_CH_HYSTERESIS(channel);
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	ret = ad7091r8_spi_reg_read(dev, reg, &data);
+	if (ret)
+		return ret;
+
+	*value |= no_os_field_get(AD7091R8_CHAN_LIMIT_MASK, data) << 3;
+
+	return 0;
+}
+
+/**
+ * @brief Initiate a software reset or hardware reset through the RESET pin.
+ * @param dev - ad7091r8_dev device handler.
+ * @param is_software - true: Software reset
+ * 		      - false: hardware reset
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad7091r8_reset(struct ad7091r8_dev *dev, bool is_software)
+{
+	int ret;
+
+	if (!dev)
+		return -EINVAL;
+
+	if (is_software) {
+		/* Bit is cleared automatically after reset */
+		return ad7091r8_spi_write_mask(dev, AD7091R8_REG_CONF,
+					       REG_CONF_RESET_MASK,
+					       NO_OS_BIT(9));
+	} else {
+		ret = no_os_gpio_set_value(dev->gpio_reset, NO_OS_GPIO_LOW);
+		if (ret)
+			return ret;
+
+		no_os_udelay(1);
+		return no_os_gpio_set_value(dev->gpio_reset, NO_OS_GPIO_HIGH);
+	}
+}
+
+/***************************************************************************//**
+ * Initializes the communication peripheral and the initial Values for
+ * AD7092R-8 Board.
+ * @param device     - The device structure.
+ * @param init_param - The structure that contains the device initial
+ * 		       parameters.
+ *
+ * @return ret - The result of the initialization procedure.
+ *               Example: -1 - SPI peripheral was not initialized or the
+ *                             device is not present.
+ *                         0 - SPI peripheral was initialized and the
+ *                             device is present.
+*******************************************************************************/
+int ad7091r8_init(struct ad7091r8_dev **device,
+		  struct ad7091r8_init_param *init_param)
+{
+	struct ad7091r8_dev *dev;
+	int ret;
+
+	if (!device)
+		return -EINVAL;
+
+	dev = (struct ad7091r8_dev *)no_os_calloc(1, sizeof(*dev));
+	if (!dev)
+		return -ENOMEM;
+
+	ret = no_os_spi_init(&dev->spi_desc, init_param->spi_init);
+	if (ret) {
+		printf("Failed to init SPI device: %d\n\r", ret);
+		goto err_release_spi;
+	}
+
+	dev->device_id = init_param->device_id;
+
+	ret = no_os_gpio_get(&dev->gpio_convst, init_param->gpio_convst);
+	if (ret) {
+		printf("Failed to get CONVST GPIO\n\r");
+		goto err_release_convst;
+	}
+
+	ret = no_os_gpio_direction_output(dev->gpio_convst, NO_OS_GPIO_HIGH);
+	if (ret) {
+		printf("Failed to set CONVST GPIO output\n\r");
+		goto err_release_convst;
+	}
+
+	ret = no_os_gpio_get_optional(&dev->gpio_reset, init_param->gpio_reset);
+	if (!ret && dev->gpio_reset) {
+		ret = no_os_gpio_direction_output(dev->gpio_reset,
+						  NO_OS_GPIO_HIGH);
+		if (ret)
+			printf("WARNING: Failed to set RESET GPIO output\n\r");
+	}
+
+	ret = ad7091r8_reset(dev, !dev->gpio_reset);
+	if (ret)
+		printf("WARNING: Failed to RESET on power up\n\r");
```

Comment:
I would remove the printf statements and return some error code to the caller instead. This one in particular is wrong anyway in the sense that it's a very bad thing not to be able to reset the chip and still continue (no return) configuring. The chip is in an unknown state if not reset.